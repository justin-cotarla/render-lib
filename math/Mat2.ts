import { AbstractMat } from './AbstractMat'
import { Vec2 } from './Vec2'

class Mat2 extends AbstractMat<Mat2, Vec2> {
  constructor(private readonly elements: [[number, number], [number, number]]) {
    super(2, Vec2.fromArray as (elements: number[]) => Vec2)
  }

  public clone = (): Mat2 => {
    return new Mat2([
      [this[0][0], this[0][1]],
      [this[1][0], this[1][1]],
    ])
  }

  get 0(): number[] {
    return this.elements[0]
  }

  get 1(): number[] {
    return this.elements[1]
  }

  set 0(value: [number, number]) {
    this.elements[0] = value
  }

  set 1(value: [number, number]) {
    this.elements[1] = value
  }

  public static fromRows = (rows: [Vec2, Vec2]): Mat2 => {
    return new Mat2([rows[0].toArray(), rows[1].toArray()])
  }

  public static fromCols = (cols: [Vec2, Vec2]): Mat2 => {
    return Mat2.fromRows(cols).transpose()
  }

  protected minor = (x: number, y: number): number => {
    return this[x][y]
  }

  public determinant = (): number => {
    return this[0][0] * this[1][1] - this[0][1] * this[1][0]
  }
}

export { Mat2 }
