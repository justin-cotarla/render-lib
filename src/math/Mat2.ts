import { AbstractMat } from './AbstractMat'
import { Vec2 } from './Vec2'

export type Mat2ElementTuple = [[number, number], [number, number]]

export class Mat2 extends AbstractMat<Mat2, Vec2> {
  static ARITY = 2

  constructor(private readonly elements: Mat2ElementTuple) {
    super(Mat2.ARITY, Vec2.fromArray as (elements: number[]) => Vec2)
  }

  public static fromRows = (rows: [Vec2, Vec2]): Mat2 => {
    return new Mat2([rows[0].toArray(), rows[1].toArray()])
  }

  public static fromCols = (cols: [Vec2, Vec2]): Mat2 => {
    return Mat2.fromRows(cols).transpose()
  }

  public static identity = (): Mat2 => {
    return new Mat2(this.identityElements(Mat2.ARITY) as Mat2ElementTuple)
  }

  public clone = (): Mat2 => {
    return new Mat2([
      [this[0][0], this[0][1]],
      [this[1][0], this[1][1]],
    ])
  }

  get 0(): number[] {
    return this.elements[0]
  }

  get 1(): number[] {
    return this.elements[1]
  }

  set 0(value: Mat2ElementTuple[0]) {
    this.elements[0] = value
  }

  set 1(value: Mat2ElementTuple[0]) {
    this.elements[1] = value
  }

  public determinant = (): number => {
    return this[0][0] * this[1][1] - this[0][1] * this[1][0]
  }

  protected minor = (x: number, y: number): number => {
    return this.minorElements(x, y)[0][0]
  }

  protected classicalAdjoint = (): Mat2 => {
    return new Mat2(this.classicalAdjointElements() as Mat2ElementTuple)
  }
}
