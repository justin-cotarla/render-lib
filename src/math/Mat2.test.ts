import { Mat2 } from './Mat2'

describe('Mat2', () => {
  describe('clone', () => {
    it('clones the matrix', () => {
      const matrix = new Mat2([1, 2, 3, 4])

      const clone = matrix.clone()

      expect(matrix.data).not.toBe(clone.data)
      expect(matrix.data).toEqual(clone.data)
    })
  })

  describe('identity', () => {
    it('returns the identity matrix', () => {
      const matrix = new Mat2().identity()
      expect(matrix.data).toEqual([1, 0, 0, 1])
    })
  })

  describe('toString', () => {
    it('prints its value', () => {
      const matrix = new Mat2([1, 2, 3, 4])
      expect(matrix.toString()).toMatchSnapshot()
    })
  })

  describe('add', () => {
    it('computes the addition of two matrices', () => {
      const matrix = new Mat2([1, 2, 3, 4]).add(new Mat2([10, 20, 30, 40]))
      expect(matrix.data).toEqual([11, 22, 33, 44])
    })
  })

  describe('subtract', () => {
    it('computes the subtraction of two matrices', () => {
      const matrix = new Mat2([1, 2, 3, 4]).subtract(new Mat2([10, 20, 30, 40]))
      expect(matrix.data).toEqual([-9, -18, -27, -36])
    })
  })

  describe('scale', () => {
    it('computes the multiplication of a matrix by a scalar', () => {
      const matrix = new Mat2([1, 2, 3, 4]).scale(10)
      expect(matrix.data).toEqual([10, 20, 30, 40])
    })
  })

  describe('multiply', () => {
    it('computes the multiplication of two matrices', () => {
      const matrix = new Mat2([1, 2, 3, 4]).multiply(new Mat2([10, 20, 30, 40]))
      expect(matrix.data).toEqual([70, 100, 150, 220])
    })
  })

  describe('determinant', () => {
    it('computes the determinant', () => {
      const matrix = new Mat2([1, 2, 3, 4])

      expect(matrix.determinant()).toBe(-2)
    })
  })

  describe('inverse', () => {
    it('computes the inverse', () => {
      const matrix = new Mat2([4, 3, 3, 2])

      expect(matrix.inverse().data).toEqual([-2, 3, 3, -4])
    })

    it('throws an error if the matrix is not invertible', () => {
      const matrix = new Mat2([4, 3, 0, 0])

      expect(() => matrix.inverse()).toThrow()
    })
  })
})
