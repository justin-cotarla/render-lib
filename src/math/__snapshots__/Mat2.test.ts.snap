// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mat2 toString prints its value 1`] = `
"[
 1, 2
 3, 4
]"
`;
